### ISSUE TITLE
Cursor session id (none) is not the same as the operation context's  - release 5.0.15 broken

### ISSUE THREAD

**Issue by simllll** [Automattic/mongoose#6357]
bug with the latest mongoose (guess the underyling problem lies in mongodb-core though).

**What is the current behavior?**
After upgrading to 5.0.15 no queries are running anymore:
```
Unhandled rejection MongoError: Cursor session id (none) is not the same as the operation context's session id (f6a28f6f-7dfa-424d-8691-fd40b12733ee - 0rbp3hXFaGQzW/eEz2QZrZ7et36CzL3bnxgnuRX9rEo=)
    at /home/simon/Dev/Web/hokify-server/node_modules/mongodb-core/lib/connection/pool.js:598:61
    at authenticateStragglers (/home/simon/Dev/Web/hokify-server/node_modules/mongodb-core/lib/connection/pool.js:516:16)
    at Connection.messageHandler (/home/simon/Dev/Web/hokify-server/node_modules/mongodb-core/lib/connection/pool.js:552:5)
    at emitMessageHandler (/home/simon/Dev/Web/hokify-server/node_modules/mongodb-core/lib/connection/connection.js:309:10)
    at Socket.<anonymous> (/home/simon/Dev/Web/hokify-server/node_modules/mongodb-core/lib/connection/connection.js:452:17)
    at Socket.emit (events.js:159:13)
    at addChunk (_stream_readable.js:265:12)
    at readableAddChunk (_stream_readable.js:252:11)
    at Socket.Readable.push (_stream_readable.js:209:10)
```

**If the current behavior is a bug, please provide the steps to reproduce.**
Just upgrade it, didn't change anything else.

**What is the expected behavior?**
Work the same way like it does with 5.0.14. Downgrading to .14 solves all problems instantely.

**Please mention your node.js, mongoose and MongoDB version.**
node v9.3.0
mongoose 5.0.15
mongodb 3.6.4

**Comment by lineus** [comment]
@simllll I'd like to dig into this, can you provide an example script that works for you on 5.0.14 but not on 5.0.15?

**Comment by simllll** [comment]
@lineus any idea how I get a backtrace to track this a bit more down?

I tried several approaches to extract the issue, but can't find it. But if I start my application the error appears, unfortuantely I cannot see what triggered it.

**Comment by lineus** [comment]
@simllll I would start the node process with `# node --inspect-brk myapp.js` and then attach a debugger ( ie. chrome dev tools or vscode ). You can then use the debugger to step through the app until it breaks. It can be tedious, but that's the only way I know of to really dig in to it.

**Comment by simllll** [comment]
I kind of narrowed it down already, and unlikely my first guess is that this is also related with the connection issue I experienced awhile ago (https://jira.mongodb.org/plugins/servlet/mobile#issue/NODE-1340).

I enabled debug log for mongodb-native and that's the output when I do 7 very simple queries in parallel (get entries from a table limited by one key .. something like translate.find({'lang': 'de'})).

[debug_all.log](https://github.com/Automattic/mongoose/files/1927245/debug_all.log)
[debug_tofirsterror.log](https://github.com/Automattic/mongoose/files/1927246/debug_tofirsterror.log)

But now the strange part: When I do the queries with a simple test script, everything works out. The situation when it's broken is when this function:
```
function __readLocale(locale, async) {
    if (async) {
        return DBController.dbConnection
            .then(() => Translation.find({lang: locale})
                .select('key phrase')
                .lean()
                .then((translations) => {
                    if (!locales[locale]) locales[locale] = {};
                    translations.forEach((t) => {
                        locales[locale][t.key] = t.phrase;
                    });
                    return locales[locale];
                }));
    }

    return locales[locale];
}
```
is called via module (it's a fork of "i18n"). Basically there is no difference, but somehow it fails then?

Test script which is working though:
```
const localesA = ['de', 'hu', 'en', 'hr', 'cs', 'es', 'it'];
            const locales = {};
            localesA.forEach(locale => {
                promises.push(Translation.find({lang: locale})
                    .select('key phrase')
                    .lean()
                    .then((translations) => {
                        if (!locales[locale]) locales[locale] = {};
                        translations.forEach((t) => {
                            locales[locale][t.key] = t.phrase;
                        });
                        return locales[locale];
                    }));
            });
```

Can you read something from the attached logs? 

Thanks
Simon

**Comment by vkarpov15** [comment]
As a workaround, try passing the `skipSessions` option to the failing query.

```javascript
MyModel.find({}, null, { skipSessions: true });
```

The MongoDB driver implicitly creates sessions if you're connected to MongoDB 3.6 running WT, that option will turn off that behavior. Should help with this issue.

**Comment by simllll** [comment]
They fixed the issue already upstream. Unfortunately no new release of mongodb-core yet... So far I go with older mongoose ;)

Thanks for the workaround.

Simon

**Comment by vkarpov15** [comment]
I'll keep this open to track, glad they fixed the issue :+1: I got bit today by implicit sessions as well, got some good feedback for the MongoDB driver team to improve them

**Comment by simllll** [comment]
New mongodb-core got released today, as soon as mongoose upgrades to 5.0.18 tthe issues should be done/gone!

Thanks
